### PR TITLE
feat: Generate validations for start and end range fields

### PR DESCRIPTION
**Pull request by aababilov**
Some tables have start and end range fields that must be in order,
e.g., `feed_info.txt` has `feed_start_date` and `feed_end_date`.

We introduce a new annotation `@EndRange` that generates a
SingleEntityValidator automatically. That reduces the amount of manually
written code and unit tests.

Example.

```
@GtfsTable(value = "feed_info.txt", singleRow = true)
public interface GtfsFeedInfoSchema extends GtfsEntity {
  @EndRange(field = "feed_end_date", allowEqual = true)
  GtfsDate feedStartDate();

  GtfsDate feedEndDate();
}
```

Generated code.

```
@Generated
@GtfsValidator
public final class GtfsFeedInfoEndRangeValidator extends SingleEntityValidator<GtfsFeedInfo> {
  @Override
  public void validate(GtfsFeedInfo entity, NoticeContainer noticeContainer) {
    if (entity.hasFeedStartDate() && entity.hasFeedEndDate()) {
      if (entity.feedStartDate().isAfter(entity.feedEndDate())) {
        noticeContainer.addValidationNotice(
            new StartAndEndRangeOutOfOrderNotice(
                GtfsFeedInfoTableLoader.FILENAME,
                entity.csvRowNumber(),
                entity.feedStartDate().toString(),
                entity.feedEndDate().toString()));
      }
    }
  }
}
```

### CODE REVIEW

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidator.java [hunk 1]
```diff
@@ -20,19 +20,18 @@
...
 import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
-import org.mobilitydata.gtfsvalidator.notice.StartAndEndDateOutOfOrderNotice;
 import org.mobilitydata.gtfsvalidator.table.GtfsFeedInfo;
 import org.mobilitydata.gtfsvalidator.table.GtfsFeedInfoTableContainer;
 
 /**
- * Validates 3 rules: 1) start_date &lt;= end_date for all rows in "feed_info.txt" 2)
- * feed_info.start_date is provided if feed_info.end_date is provided 3) feed_info.end_date is
- * provided if feed_info.start_date is provided.
+ * Validates that if one of {@code (start_date, end_date)} fields is provided for {@code
```
Comment (lionel-nj) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidator.java:27]: Thanks for clarifying this

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+  private static List<ValidationNotice> generateNotices(GtfsFrequency frequency) {
+    NoticeContainer noticeContainer = new NoticeContainer();
+    GtfsFrequencyEndRangeValidator validator = new GtfsFrequencyEndRangeValidator();
+    validator.validate(frequency, noticeContainer);
+    return noticeContainer.getValidationNotices();
+  }
+
+  @Test
+  public void inOrder_yieldsNoNotice() {
```
Comment (lionel-nj) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:46]:
```suggestion
  public void timesInOrderShouldNotYieldsNotice() {
```

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                new GtfsFrequency.Builder()
+                    .setCsvRowNumber(2)
+                    .setStartTime(GtfsTime.fromString("10:00:00"))
+                    .setEndTime(GtfsTime.fromString("11:00:00"))
+                    .build()))
+        .isEmpty();
+  }
+
+  @Test
+  public void outOfOrder_yieldsNotice() {
```
Comment (lionel-nj) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:58]:
```suggestion
  public void timesOutOfOrderShouldYieldsNotice() {
```

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                    .setStartTime(GtfsTime.fromString("11:00:00"))
+                    .setEndTime(GtfsTime.fromString("10:00:00"))
+                    .build()))
+        .containsExactly(
+            new StartAndEndRangeOutOfOrderNotice(
+                GtfsFrequencyTableLoader.FILENAME, 2, "11:00:00", "10:00:00"));
+  }
+
+  @Test
+  public void equal_yieldsNotice() {
```
Comment (lionel-nj) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:72]:
```suggestion
  public void equaTimesShouldYieldNotice() {
```

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                    .setStartTime(GtfsTime.fromString("10:00:00"))
+                    .setEndTime(GtfsTime.fromString("10:00:00"))
+                    .build()))
+        .containsExactly(
+            new StartAndEndRangeEqualNotice(
+                GtfsFrequencyTableLoader.FILENAME, 2, "10:00:00", "10:00:00"));
+  }
+
+  @Test
+  public void fieldMissing_yieldsNoNotice() {
```
Comment (lionel-nj) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:86]:
```suggestion
  public void missingFieldShouldNotYieldNotice() {
```

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopTimeArrivalAndDepartureTimeValidatorTest.java [hunk 6]
```diff
@@ -87,34 +86,6 @@ public void departureTimeAfterArrivalTimeShouldNotGenerateNotice() {
     assertThat(noticeContainer.getValidationNotices()).isEmpty();
   }
 
-  @Test
-  public void departureTimeBeforeArrivalTimeShouldGenerateNotice() {
```
Comment (lionel-nj) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopTimeArrivalAndDepartureTimeValidatorTest.java:91]: `departureTimeAfterArrivalTimeShouldNotGenerateNotice` the test just above this one should be removed too.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidatorTest.java [hunk 7]
```diff
@@ -63,31 +62,6 @@ public void startDateBeforeEndDateShouldNotGenerateNotice() {
     assertThat(noticeContainer.getValidationNotices().isEmpty());
   }
 
-  @Test
-  public void startDateAfterEndDateShouldGenerateNotice() {
```
Comment (lionel-nj) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidatorTest.java:67]: `startDateBeforeEndDateShouldNotGenerateNotice` the test just above this one should be removed too.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFieldDescriptor.java [hunk 8]
```diff
@@ -85,6 +87,10 @@
 
     public abstract Builder setDefaultValue(String value);
 
+    public abstract Builder setEndRange(Optional<EndRangeDescriptor> value);
```
Comment (lionel-nj) [inline processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFieldDescriptor.java:90]: Is this to avoid throwing NPE?

**Review by lionel-nj — COMMENTED** [review]
Thanks @aababilov! That certainly reduces the amount of code for date and time range validation. Just a few suggestions in-line: what do you think about renaming some test methods? And a question regarding the use of `Optional`  in `setEndRange()` method.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFieldDescriptor.java (hunk 8 above)
Comment (aababilov) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFieldDescriptor.java:90]: No, this is a convention of AutoValue. This way you can clean up the field: `setEndRange(Optional.empty())`.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidator.java (hunk 1 above)
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidator.java:27]: My pleasure!

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java (hunk 3 above)
Comment (aababilov) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:58]: I am following the Google style guide here. Words like "should" are redundant. Underscores make test names more readable: they separate the actions taken from the expected outcome.

The Google style guide says:

A test's name should summarize the behavior it is testing. A good name describes both the actions that are being taken on a system and the expected outcome. Test names will sometimes include additional information like the state of the system or its environment before taking action on it.

* transferMoney_deductsFromSource
* transferMoney_insufficientFunds_throwsException
* multiply_twoNegativeValues_returnsPositiveValue
* isInputValid_conversionSucceeds

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java (hunk 4 above)
Comment (aababilov) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:72]: See above.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java (hunk 5 above)
Comment (aababilov) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:86]: See above.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidatorTest.java (hunk 7 above)
Comment (aababilov) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/validator/FeedServiceDateValidatorTest.java:67]: Good catch!

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopTimeArrivalAndDepartureTimeValidatorTest.java (hunk 6 above)
Comment (aababilov) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopTimeArrivalAndDepartureTimeValidatorTest.java:91]: Good catch!

**Comment by aababilov** [comment]
Thank you for review!

Yes, my goal is to remove the amount of production code and tests. Annotations are so powerful when it comes to typical checks.

I would like to keep the test names and actually establish them as a new convention.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java (hunk 3 above)
Comment (lionel-nj) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:58]: Makes sense to me! @barbeau?

**Review by lionel-nj — APPROVED** [review]
LGTM, I'll let you resolve the conflicts @aababilov. Thanks for this.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java (hunk 3 above)
Comment (barbeau) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:58]: I agree that the underscore pattern is easier to read, and is allowed by the Google Java Style Guide that we've previously agreed upon:
https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names

>Underscores may appear in JUnit test method names to separate logical components of the name, with each component written in lowerCamelCase. One typical pattern is <methodUnderTest>_<state>, for example pop_emptyStack. There is no One Correct Way to name test methods.

I would propose we follow this pattern going forward. We can refactor existing tests as we touch them.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import org.junit.runners.JUnit4;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.notice.StartAndEndRangeEqualNotice;
+import org.mobilitydata.gtfsvalidator.notice.StartAndEndRangeOutOfOrderNotice;
+import org.mobilitydata.gtfsvalidator.notice.ValidationNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsFrequency;
+import org.mobilitydata.gtfsvalidator.table.GtfsFrequencyTableLoader;
+import org.mobilitydata.gtfsvalidator.type.GtfsTime;
+
+// FIXME: Cover autogenerated end-range validators by unit tests in "processor".
```
Comment (barbeau) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsFrequencyEndRangeValidatorTest.java:33]: @aababilov Could you open an issue on this GitHub repo describing what needs to be done to resolve this FIXME?

**Review by barbeau — APPROVED** [review]
🚀  Thanks @aababilov, this is very nice! We can merge as-is, although please see note about opening an issue regarding the `FIXME`.

**Comment by barbeau** [comment]
@aababilov Sorry, actually can't merge as-is - we can merge after you resolve conflicts.

**Comment by aababilov** [comment]
Thanks for review! I resolved the conflicts.

**Comment by aababilov** [comment]
Issue for the tests: https://github.com/MobilityData/gtfs-validator/issues/818

**Pull request merged**